### PR TITLE
add start of local partial moran statistics

### PR DESCRIPTION
**Pull request by ljwolf**
This adds my partial and auxiliary-regressive local Moran statistics forthcoming in the annals. This needs

- [ ] tests
- [ ] example notebook

### CODE REVIEW

**Comment by ljwolf** [comment]
now that the paper has hit the annals, I need to get this merged. I've resumed stubbing the tests out, and they should be simple to finalize. I'm hopeful that we can get this released in the next ESDA iteration. 

I think that, after writing another two class in this frame, it would help to do a similar RFC on an api for ESDA-style interfaces that we can agree upon for the future. I think we need to start developing a more consistent interface, with attributes that have less specificity to the implementation of the class (p_sim vs. p_norm; I vs. associations_ vs. Gi, etc.) 

I hope to have this RFC written for the development meeting next week.

File: esda/moran_local_mv.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+                    partial regressive contribution of that covariate, with
+                    the first column indicating the part-regressive outcome
+                    and the second indicating the part-regressive design.
+                    The partial regression matrix starts at zero, so
+                    partials_[:,:,0] corresponds to the partial regression
+                    describing the relationship between y and Wy.
+        """
+        self._mvquads = mvquads
+        y = np.asarray(y).reshape(-1, 1)
+        W.transform = "r"
```
Comment (martinfleis) [inline esda/moran_local_mv.py:82]: Can you ensure support of Graph? This still assumes W only.

Comment (ljwolf) [reply esda/moran_local_mv.py:84]: Yes! I think I've done this with the following check: 

```python
        if hasattr(W, "to_W"):
            W = W.to_W()
```

Comment (martinfleis) [reply esda/moran_local_mv.py:84]: The rest of esda is Graph-native now, so there's no conversion happening. If we can keep it that way, I'd be happier but understand that it may be a larger undertaking. It is up to you.

**Comment by ljwolf** [comment]
Not sure why the CI is failing on the `pygrio+geodatasets` read?

**Comment by ljwolf** [comment]
OK, it looks like the `geodatasets` download is malformed, so the `pyogrio` or `fiona` read fails for one worker. When that happens, the process fails for `test_moran_local_mv.py`, and we get the discrepancy between the runners. Maybe a fixture will solve it?

**Comment by martinfleis** [comment]
> OK, it looks like the geodatasets download is malformed


the issue is that multiple workers are trying to donwload and save the same file at the same time. We need to pull the data ahead of time to make that work as we do in lib https://github.com/pysal/libpysal/blob/bef1e81aa35c9372e91a142e9fa5c8c0c44c55ee/.github/workflows/unittests.yml#L64-L76

**Comment by ljwolf** [comment]
I think that's one useful solution! I think I would rather do this as a `pytest.fixture`, however, since that allows us to [specify the scope of the fixture](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session), and for `pytest` then to understand that scope. 

```
:param scope:
    The scope for which this fixture is shared; one of ``"function"``
    (default), ``"class"``, ``"module"``, ``"package"`` or ``"session"``.

    This parameter may also be a callable which receives ``(fixture_name, config)``
    as parameters, and must return a ``str`` with one of the values mentioned above.

    See :ref:`dynamic scope` in the docs for more information.
```

Using the module-level fixture, this passes now. 
We should probably define these datasets in `libpysal/tests/__init__.py` as pytest fixtures at the package level.

**Comment by martinfleis** [comment]
Yeah, that is one option. I am not a big fan of it as tests with a ton of fixtures are pain to debug if you want to re-run those in ipython or a notebook but it is one way of solving this.

**Comment by ljwolf** [comment]
OK! I'm fine with that, too. I like fixtures, since they can encapsulate both the parametrize step and can compose/depend correctly across parallel jobs. So, fine to keep what's in `libpysal` anyway, if we're cool using fixtures here.

File: esda/moran_local_mv.py (hunk 1 above)
Comment (ljwolf) [reply esda/moran_local_mv.py]: This should now be done. 

One thing of note... if we were to swap `w.sparse * x` for `w.sparse @ x` in `libpysal.weights.spatial_lag.lag_spatial()`, then all instances of `lag_spatial` would work on Graph, too, while preserving backwards compatibility...

**Comment by martinfleis** [comment]
> So, fine to keep what's in libpysal anyway, if we're cool using fixtures here.

I'm fine with either, I just find working with fixtures a bit opaque occasionally and a long sequence of them is super pain to replicate in a notebook.

File: esda/moran_local_mv.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+try:
+    from tqdm import tqdm
+except ImportError:
+
+    def tqdm(x, **kwargs):
+        return x
+
+
+class Partial_Moran_Local(object):
```
[the remark this answers is not in the document]
Comment (ljwolf) [reply esda/moran_local_mv.py:14]: It's the current api... I'm moving towards the lowercase moran_local for the new api.

**Comment by ljwolf** [comment]
OK, this should be ready to review in full/merge. 

This commits the estimators in the same convention as the existing `Moran` estimators, but uses a slightly different convention for describing the statistic outputs. I would like to use that as the basis for new classes in `esda`, and hope to have that by the next dev meeting (again 😄 )

**Comment by martinfleis** [comment]
> Tests were all passing in [29b231](https://github.com/pysal/esda/pull/279/commits/29b2312b13f15077171df0a9d693ab320d34dab2), and new failures are in unrelated code in topo. Not sure the origin yet.

Incompatibility of shapely with numpy 2.1 released yesterday.

**Comment by sjsrey** [comment]
> OK, this should be ready to review in full/merge. The example notebook is complete.
> 
> This commits the estimators in the same convention as the existing `Moran` estimators, but uses a slightly different convention for describing the statistic outputs. I would like to use that as the basis for new classes in `esda`, and hope to have that by the next dev meeting (again 😄 )



I'm looking foward to the discussion. One question is how closely the plan is to track scikit-learn conventions, as the implementation in [moran_local_mv.py](https://github.com/ljwolf/esda/blob/mvlisa/esda/moran_local_mv.py) does not follow the "parameters with trailing underscore `_` are not to be set inside the `__init__` method" rule.[^sk_]

A point for discussion at the dev meeting.

This is looking really nice!




[^sk_]: https://github.com/rasbt/python-machine-learning-book/blob/master/faq/underscore-convention.md

**Comment by ljwolf** [comment]
I've updated the API to follow that resource (I think?). Only configuration options are set in the init. 

When I mentioned using this PR as a prototype for a new spatial statistics API, I meant exactly this!

One thing I'd like feedback here is *how much data to cache*. Here, for example, in the `Partial_Moran_Local()` class, I'm caching three `(n_samples, n_features)` matrices (`self.D`, `self.R`, `self._partials_`), plus the `(n_samples,n_permutations,n_variables)` matrix `self._rlmos_`). 

Generally, `sklearn` classes avoid caching the input data if possible, but this means I'd need to remove the properties and just assign directly the attributes we want to keep. For naming, I'm thinking:

- `association_`/`associations_`: the global/local measure of spatial association that the statistic expresses.
- `labels_`: the "label", if any, for common local statistics. Here, I'm thinking just the quadrant for Moran-style local covariance statistics, or hotspot/coldspot for the Getis-Ord statistics.
- `significance_`/`significances_`: local p-values that are generated according to an option passed in the `__init__()`. These will generally be *calculated* in the `fit()` object, if we want to avoid caching data on the object itself. 
- `reference_distribution_`: if permutation inference is used and the random values are cached, this should be where the simulations are stored. 
- `partials_`: for any cross-product style statistic, this should record the two vectors that are crossed before calculating a global statistic, or elementwise-multiplied before calculating a local statistic. This is to facilitate visualization later, as well as allow for new labelling in a `.transform()` method.

**Comment by jGaboardi** [comment]
@ljwolf Did you verify the docs build locally?

**Comment by ljwolf** [comment]
@jGaboardi I've done so now. We talked about the following in the call:

- [x] switch to `CamelCase`
- [x] move to singular nouns (`association_`, `significance_`, `reference_distribution_`) when these would look strange for univariate statistics, but keep plurals (`partials_`, `labels_`) for others
- [x] add to `__init__.py` 
- [x] render docs and update gallery

But, I've now started moving the implementation to new significance engine from #281, so I will need to update the tests. Then, this is ready to ship!

**Comment by ljwolf** [comment]
Yes please!
